### PR TITLE
FIP-53 : d3pool farm on Convex

### PR DESCRIPTION
**Pull request by eswak**
# FIP-53
This proposal mints 50M FEI to deposit in the d3pool on Curve.
The Curve LP tokens are staked on Convex, to earn CRV and CVX rewards, as well as trading fees.
FEI liquidity against FRAX and alUSD will be greatly enhanced.

# Curve and Convex PCV deposits
This branch introduces two new PCVDeposits : one for Curve plain pools, and one for Convex. They are detailed below.

## Curve PCV deposit (plain pool)
This Curve deposit works with plain pools (pool composed of 3 stable assets, like the 3pool or d3pool).
It will `deposit()` all the ERC20s held of the 3 assets of the pool, to get Curve LP tokens.

The Curve LP tokens stay on the contract. You can move them to a Convex deposit by using `withdrawERC20()`.

The `withdraw()` call removes only one token (FEI if present, or the first token in the pool at index 0). To choose which token to withdraw, use `withdrawOneToken()`. These withdrawals might have some slippage if a large amount is withdrawn and makes the pool imbalanced. To burn all LPs and exit the LP position by withdrawing all 3 tokens of the pool in proportions that are present in the pool, and not get any slippage (besides withdrawal fees), call `exitPool()`.

Tests:
image

## Convex deposit (generic)
The Convex deposit takes a given kind of Curve LP tokens, `deposit()` and stake them on Convex, and can then claim CRV and CVC rewards. A call to `withdraw()` claims all unclaimed rewards, and pulls back the Curve LP tokens on the Convex deposit.

image

### CODE REVIEW

File: proposals/description/fip_53.ts
Comment (ElliotFriedman) [inline proposals/description/fip_53.ts:48]: This proposal is incomplete. It should also deploy a new PCVDeposit to add this 50m FEI to protocol controlled FEI and add that PCVDeposit to the CR Oracle.

**Review by ElliotFriedman — DISMISSED** [review]

**Comment by Joeysantoro** [comment]
How do we manage the CR oracle here?

**Comment by eswak** [comment]
@Joeysantoro @ElliotFriedman 

I'm thinking we add 50M protocol fei to the StaticPCVDepositWrapper through OA, because eventually we can exit the pool by removing ~49.8M FEI. Or we add 16.67M Protocol Fei and 33.33M$ PCV (ideal balance of the Curve pool : we'll own 16.67M FRAX and 16.67M alUSD).

A new PCVDeposit means new code that needs review & maybe even audit. I think for now, we can move straight from the Timelock. It looks like we're going to interface more with Curve/Convex, so I'll work on proper PCVDeposit contracts, but we can move the LP tokens there later.

**Comment by eswak** [comment]
Nevermind, I'm writing a simple Convex PCVDeposit, should be quick (& cleaner).

File: proposals/description/fip_53.ts
Comment (eswak) [reply proposals/description/fip_53.ts:48]: Added a new PCVDeposit to do it, and updated the proposal to track the new deposit through CR Oracle

**Review by Joeysantoro — CHANGES_REQUESTED** [review]
This code is too complicated imo. I think it should take in d3 tokens and output d3 tokens. Just a basic staking wrapper. We can add some code that can wrap and unwrap d3 as a separate contract

File: contracts/pcv/curve/CurvePCVDepositPlainPool.sol
Comment (ElliotFriedman) [inline contracts/pcv/curve/CurvePCVDepositPlainPool.sol:20]: This should be a type of `ICurveStableSwap3 ` instead of address as many other places in the codebase it is casted to a `ICurvePool` or `ICurveStableSwap3` and the ICurvePool interface is inherited in the ICurveStableSwap3 interface.

**Review by Joeysantoro — COMMENTED** [review]
Solidity LGTM pending review from Curve and/or Convex team + another review of the tests and deploy script

File: contracts/pcv/curve/ICurveStableSwap3.sol
Comment (ElliotFriedman) [inline contracts/pcv/curve/ICurveStableSwap3.sol:11]: format

File: contracts/pcv/curve/ICurvePool.sol
Comment (ElliotFriedman) [inline contracts/pcv/curve/ICurvePool.sol:20]: format

File: test/unit/pcv/ConvexPCVDeposit.test.ts
Comment (ElliotFriedman) [inline test/unit/pcv/ConvexPCVDeposit.test.ts:77]: add init tests

File: test/unit/pcv/CurvePCVDepositPlainPool.test.ts
Comment (ElliotFriedman) [inline test/unit/pcv/CurvePCVDepositPlainPool.test.ts:60]: add init tests

**Comment by eswak** [comment]
Improved the pull-request's description to better document what has been done.

I also took your 2nd review into account @ElliotFriedman, thanks 👍 

I have shared the code with some Convex team members to get a sanity check of how we interact with their contracts.

**Review by Joeysantoro — COMMENTED** [review]
I'd like to see at least one integration test going through entire flow. Otherwise LGTM

File: test/integration/tests/fip-53.ts
Comment (Joeysantoro) [inline test/integration/tests/fip-53.ts:87]: fast forward time to check CRV and CVX balances?

**Review by Joeysantoro — DISMISSED** [review]

**Review by Joeysantoro — APPROVED** [review]

**Pull request merged**
